Adding PollerFlux::create to support ActivationOp variant that returns PollResponse

anuchandy opened the pull request.

srnagar commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -89,17 +90,94 @@ public PollerFlux(Duration defaultPollInterval,
+        }
+        this.defaultPollInterval = defaultPollInterval;
+        this.activationOperation = null;
+        this.activationOperationEx = Objects.requireNonNull(activationOperation,
```
nit: rename this to `activationOperationWithResponse` to make it clear that this is different from `activationOperation` which doesn't have the PollResponse.

anuchandy replied:
I'm trying to get rid of storing activationOperation in class level, just updated the code to move all activation logic to its own inner class.

srnagar commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -89,17 +90,94 @@ public PollerFlux(Duration defaultPollInterval,
     public void subscribe(CoreSubscriber<? super AsyncPollResponse<T, U>> actual) {
         this.oneTimeActivationMono
-            .flatMapMany(ignored -> pollingLoop())
+            .flatMapMany(ignored -> {
```
Would be good to add a unit test case for this change.

anuchandy replied:
tests added

srnagar commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -129,34 +207,43 @@ public void subscribe(CoreSubscriber<? super AsyncPollResponse<T, U>> actual) {
             }
             if (this.guardActivationCall.compareAndSet(this, 0, 1)) {
-                final Mono<T> activationMono;
+                final boolean isEx = this.activationOperationEx != null;
```
same here - maybe name this as `isActivationWithResponse`

anuchandy commented:
@srnagar  - I just restructured the code instead of tweaking existing one by adding if/else and casting etc. I felt it started loosing readability. :(

The updates are:

1. The whole concept of one-time-activation is in its own inner class.
2. No two private vars to store activationOp references.
2. Replaced AtomicFieldUpdater with AtomicBoolean.
3. Removed un-necessary compareAndSet.

I'm adding a test for factory method. 

Please take a look when you get a chance.

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/DefaultSyncPoller.java:
```diff
@@ -38,10 +38,7 @@
-     *                            this parameter is required, if there is no specific activation work to be
-     *                            done then invocation should return Mono.empty(), this operation will be called
-     *                            with a new {@link PollingContext}.
+     * @param syncActivationOperation the activation function upon call synchronously execute the activation operation.
```
Can you please improve this wording - I'm not entirely sure what it means.

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -87,19 +82,103 @@ public PollerFlux(Duration defaultPollInterval,
+     * PollerFlux obtained from this factory method uses an activationOperation which returns a Mono that
+     * emits {@link PollResponse}, which holds the result.
+     * The PollerFlux created from constructor uses an activationOperation which returns a Mono that directly
+     * emits result.
```
If you're meaning to contrast this method against the constructor, be more explicit. "This create method differs from the PollerFlux constructor in that the constructor uses an ...."

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -87,19 +82,103 @@ public PollerFlux(Duration defaultPollInterval,
+     * The PollerFlux created from constructor uses an activationOperation which returns a Mono that directly
+     * emits result.
+     * Since the first variant of PollerFlux has access to the response, it can skip the polling loop if the
+     * response indicate that LRO is completed. Whereas the second PollerFlux variant calls pollFunction at
```
Avoid acronyms, e.g, "that the long-running operation"

JonathanGiles replied:
Similarly, I'm not entirely clear what you mean by 'first variant' and 'second variant'. Try to be more explicit to avoid any confusion.

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -87,19 +82,103 @@ public PollerFlux(Duration defaultPollInterval,
+     * @param activationOperation the activation operation to be invoked at most once across all subscriptions,
+     *                            this parameter is required, if there is no specific activation work to be
+     *                            done then invocation should return Mono.empty(), this operation will be called
+     *                            with a new {@link PollingContext}.
```
Don't worry about this kind of indentation - checkstyle should be warning against this too - please check if checkstyle passes.

JonathanGiles replied:
This is a very long, single sentence. Break it down into a few sentences.

anuchandy commented:
Thanks @JonathanGiles, will cleanup javadoc and get back.

srnagar commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -108,61 +187,12 @@ public void subscribe(CoreSubscriber<? super AsyncPollResponse<T, U>> actual) {
     public SyncPoller<T, U> getSyncPoller() {
         return new DefaultSyncPoller<>(this.defaultPollInterval,
-                this.activationOperation,
+                this.syncActivationOperation,
```
Here you can do `ctx -> this.activationOperation.apply(ctx).block()` instead of eagerly creating `syncActivationOperation`

anuchandy replied:
We could do it in previous version but not anymore since we don't have two class members defined to store variations of activationOperations

anuchandy replied:
@srnagar - or we could re-introduces those two member vars and use it for SyncPoller is created. What do you think?

srnagar replied:
@anuchandy I guess re-introducing two variables will make it less readable. We can leave it as-is for now.

anuchandy replied:
Ok, sounds good.

anuchandy commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -87,19 +82,103 @@ public PollerFlux(Duration defaultPollInterval,
+            // mapper
+            activationResult -> new PollResponse<>(LongRunningOperationStatus.NOT_STARTED, activationResult)).getMono();
+        this.syncActivationOperation =
+            cxt -> new PollResponse<>(LongRunningOperationStatus.NOT_STARTED, activationOperation.apply(cxt).block());
```
We don't block in the PollerFlux constructor/factory. We are creating  function that get passed to SyncPoller and it's gets blocked when called from SyncPoller.

anuchandy commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -87,19 +82,103 @@ public PollerFlux(Duration defaultPollInterval,
+            activationOperation,
+            // mapper
+            Function.identity()).getMono();
+        this.syncActivationOperation = cxt -> activationOperation.apply(cxt).block();
```
As mentioned in previous comment, it's function that blocked only when called from SyncPoller 

The reason for wrapping activationOpertion in Func is, we have two types of activationOperation depending on how user want to create poller - factory vs ctr.

anuchandy commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -87,19 +82,103 @@ public PollerFlux(Duration defaultPollInterval,
+     * @return PollerFlux
+     */
+    public static <T, U> PollerFlux<T, U>
+        create(Duration defaultPollInterval,
```
Will do

anuchandy commented:
@JonathanGiles - updated the JavaDoc, could you re-review?

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/DefaultSyncPoller.java:
```diff
@@ -22,7 +22,7 @@
  *
  * @param <T> The type of poll response value
- * @param <U> The type of the final result of long-running operation
+ * @param <U> The type of the final result of long running operation
```
```suggestion
 * @param <U> The type of the final result of the long running operation
```

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/DefaultSyncPoller.java:
```diff
@@ -37,42 +37,38 @@
-     *                            this parameter is required, if there is no specific activation work to be
-     *                            done then invocation should return Mono.empty(), this operation will be called
-     *                            with a new {@link PollingContext}.
+     * @param pollInterval the polling interval
```
```suggestion
     * @param pollInterval the polling interval.
```

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/DefaultSyncPoller.java:
```diff
@@ -37,42 +37,38 @@
-     *                            with a new {@link PollingContext}.
+     * @param pollInterval the polling interval
+     * @param syncActivationOperation the operation to synchronously activate (start) the long running operation,
+     *     this operation will be called with a new {@link PollingContext}.
```
This still lacks clarity to me. I don't know, as a user of this API, what exactly I should supply here. In other words, you're telling me how this is used, rather than how I use it.

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/DefaultSyncPoller.java:
```diff
@@ -37,42 +37,38 @@
-     *                        {@link PollingContext}.
+     *     is required and the operation will be called with current {@link PollingContext}.
+     * @param cancelOperation a {@link Function} that represents the operation to cancel the long running operation
+     *     if service supports cancellation, this parameter is required and if service does not support cancellation
```
Break some of these sentences up - in a few places you flow straight into "this parameter is required" when a full stop and a new sentence would improve readability.

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/DefaultSyncPoller.java:
```diff
@@ -37,42 +37,38 @@
                              Function<PollingContext<T>, Mono<U>> fetchResultOperation) {
-        Objects.requireNonNull(defaultPollInterval, "'defaultPollInterval' cannot be null.");
-        if (defaultPollInterval.compareTo(Duration.ZERO) <= 0) {
+        Objects.requireNonNull(pollInterval, "'defaultPollInterval' cannot be null.");
```
```suggestion
        Objects.requireNonNull(pollInterval, "'pollInterval' cannot be null.");
```

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -11,13 +11,13 @@
- * A Flux that simplifies the task of executing long-running operations against an Azure service.
- * A subscription to {@link PollerFlux} initiate a long-running operation and polls the status
+ * A Flux that simplifies the task of executing long running operations against an Azure service.
+ * A subscription to {@link PollerFlux} initiate a long running operation and polls the status
```
```suggestion
 * A subscription to {@link PollerFlux} initiates a long running operation and polls the status
```

JonathanGiles commented on sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java:
```diff
@@ -38,68 +38,143 @@
+    }
+
+    /**
+     * Creates PollerFlux.
```
Provide a better first sentence, as this is the most visible text in the JavaDoc.

anuchandy commented:
@JonathanGiles will send follow up Javadoc PR. Merging this to unblock mgmt work.

\\cc @yaohaizh
